File: tests/test_spaced_repetition.py
"""Tests for SM-2 spaced repetition algorithm.

These tests verify the SM-2 algorithm logic as implemented in
static/js/spaced-repetition.js by testing the same rules in Python.
They serve as a reference implementation to catch regressions.
"""

import pytest

# SM-2 constants (matching spaced-repetition.js)
DEFAULT_EASE = 2.5
MIN_EASE = 1.3

QUALITY_WRONG = 1
QUALITY_UNSURE = 3
QUALITY_GOOD = 4
QUALITY_EASY = 5


def adjust_ease(ef, quality):
    """Compute updated ease factor using SM-2 formula."""
    delta = 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)
    return max(MIN_EASE, ef + delta)


def sm2_review(reps, interval, ef, quality):
    """Apply SM-2 update rules. Returns (new_reps, new_interval, new_ef)."""
    new_ef = adjust_ease(ef, quality)

    if quality < 3:
        return (0, 1, new_ef)

    if reps == 0:
        new_interval = 1
    elif reps == 1:
        new_interval = 6
    else:
        new_interval = round(interval * new_ef)

    return (reps + 1, new_interval, new_ef)


# -------------------------------------------------------------------
# Ease factor adjustment tests
# -------------------------------------------------------------------


class TestEaseFactorAdjustment:
    def test_easy_increases_ef(self):
        new_ef = adjust_ease(DEFAULT_EASE, QUALITY_EASY)
        assert new_ef == pytest.approx(2.6)

    def test_good_keeps_ef(self):
        new_ef = adjust_ease(DEFAULT_EASE, QUALITY_GOOD)
        assert new_ef == pytest.approx(2.5)

    def test_unsure_decreases_ef(self):
        new_ef = adjust_ease(DEFAULT_EASE, QUALITY_UNSURE)
        assert new_ef == pytest.approx(2.36)

    def test_wrong_decreases_ef_more(self):
        new_ef = adjust_ease(DEFAULT_EASE, QUALITY_WRONG)
        assert new_ef == pytest.approx(1.96)

    def test_ef_minimum_clamp(self):
        """Repeated wrong answers should not drop EF below 1.3."""
        ef = MIN_EASE
        new_ef = adjust_ease(ef, QUALITY_WRONG)
        assert new_ef == MIN_EASE

    def test_ef_minimum_after_many_wrongs(self):
        ef = DEFAULT_EASE
        for _ in range(20):
            ef = adjust_ease(ef, QUALITY_WRONG)
        assert ef == MIN_EASE


# -------------------------------------------------------------------
# SM-2 interval progression tests
# -------------------------------------------------------------------


class TestIntervalProgression:
    def test_first_correct_review(self):
        reps, interval, ef = sm2_review(0, 1, DEFAULT_EASE, QUALITY_GOOD)
        assert reps == 1
        assert interval == 1

    def test_second_correct_review(self):
        reps, interval, ef = sm2_review(1, 1, DEFAULT_EASE, QUALITY_GOOD)
        assert reps == 2
        assert interval == 6

    def test_third_correct_review(self):
        reps, interval, ef = sm2_review(2, 6, DEFAULT_EASE, QUALITY_GOOD)
        assert reps == 3
        assert interval == round(6 * DEFAULT_EASE)

    def test_wrong_resets(self):
        reps, interval, ef = sm2_review(5, 100, DEFAULT_EASE, QUALITY_WRONG)
        assert reps == 0
        assert interval == 1

    def test_easy_progression(self):
        """Easy reviews increase EF, leading to longer intervals."""
        reps = 0
        interval = 1
        ef = DEFAULT_EASE

        reps, interval, ef = sm2_review(reps, interval, ef, QUALITY_EASY)
        assert reps == 1

        reps, interval, ef = sm2_review(reps, interval, ef, QUALITY_EASY)
        assert reps == 2

        reps, interval, ef = sm2_review(reps, interval, ef, QUALITY_EASY)
        assert reps == 3
        # EF has increased from 2.5 after two Easy reviews
        assert ef > DEFAULT_EASE
        assert interval > round(6 * DEFAULT_EASE)

    def test_interval_after_wrong_then_recovery(self):
        """After a wrong answer, recovery starts from scratch."""
        reps, interval, ef = sm2_review(3, 30, DEFAULT_EASE, QUALITY_WRONG)
        assert reps == 0
        assert interval == 1

        reps, interval, ef = sm2_review(reps, interval, ef, QUALITY_GOOD)
        assert reps == 1
        assert interval == 1

        reps, interval, ef = sm2_review(reps, interval, ef, QUALITY_GOOD)
        assert reps == 2
        assert interval == 6


# -------------------------------------------------------------------
# Quality value tests
# -------------------------------------------------------------------


class TestQualityValues:
    def test_quality_constants(self):
        assert QUALITY_WRONG == 1
        assert QUALITY_UNSURE == 3
        assert QUALITY_GOOD == 4
        assert QUALITY_EASY == 5

    def test_wrong_always_resets(self):
        for initial_reps in [0, 1, 5, 10]:
            reps, interval, _ = sm2_review(
                initial_reps, 100, DEFAULT_EASE, QUALITY_WRONG
            )
            assert reps == 0
            assert interval == 1

    def test_unsure_advances(self):
        """Unsure (q=3) counts as correct -- advances repetitions."""
        reps, interval, ef = sm2_review(0, 1, DEFAULT_EASE, QUALITY_UNSURE)
        assert reps == 1

    def test_unsure_decreases_ef(self):
        """Unsure reviews reduce ease factor (slower future intervals)."""
        _, _, ef = sm2_review(0, 1, DEFAULT_EASE, QUALITY_UNSURE)
        assert ef < DEFAULT_EASE

File: tests/js/test-spaced-repetition.js
/**
 * Tests for the SpacedRepetition module (SM-2 algorithm).
 */

var { describe, it } = require('node:test');
var assert = require('node:assert/strict');

require('./helpers/load-modules');

describe('SpacedRepetition', function () {
    describe('createCard', function () {
        it('returns a card with default values', function () {
            var card = SpacedRepetition.createCard('test-card');
            assert.equal(card.card_id, 'test-card');
            assert.equal(card.ease_factor, 2.5);
            assert.equal(card.interval_minutes, 1);
            assert.equal(card.repetitions, 0);
            assert.equal(card.last_quality, null);
            assert.ok(card.next_review);
        });
    });

    describe('QUALITY constants', function () {
        it('has correct quality values', function () {
            assert.equal(SpacedRepetition.QUALITY.wrong, 1);
            assert.equal(SpacedRepetition.QUALITY.unsure, 3);
            assert.equal(SpacedRepetition.QUALITY.good, 4);
            assert.equal(SpacedRepetition.QUALITY.easy, 5);
        });
    });

    describe('review', function () {
        it('resets repetitions on wrong answer', function () {
            var card = SpacedRepetition.createCard('test');
            // First correct review to build up repetitions
            card = SpacedRepetition.review(card, 4);
            assert.equal(card.repetitions, 1);

            // Wrong answer resets
            card = SpacedRepetition.review(card, 1);
            assert.equal(card.repetitions, 0);
            assert.equal(card.interval_minutes, 1);
        });

        it('advances repetitions on correct answer', function () {
            var card = SpacedRepetition.createCard('test');
            card = SpacedRepetition.review(card, 4);
            assert.equal(card.repetitions, 1);
            assert.equal(card.interval_minutes, 2);

            card = SpacedRepetition.review(card, 4);
            assert.equal(card.repetitions, 2);
            assert.equal(card.interval_minutes, 10);
        });

        it('scales interval by ease factor after rep 2', function () {
            var card = SpacedRepetition.createCard('test');
            card = SpacedRepetition.review(card, 5); // rep 1, interval 2
            card = SpacedRepetition.review(card, 5); // rep 2, interval 10
            card = SpacedRepetition.review(card, 5); // rep 3, interval = 10 * newEf
            assert.equal(card.repetitions, 3);
            // EF adjusts again on this review, so interval = round(10 * currentEf)
            assert.equal(card.interval_minutes, Math.round(10 * card.ease_factor));
        });

        it('does not mutate the input card', function () {
            var card = SpacedRepetition.createCard('test');
            var updated = SpacedRepetition.review(card, 4);
            assert.equal(card.repetitions, 0);
            assert.equal(updated.repetitions, 1);
        });

        it('adjusts ease factor downward on wrong answer', function () {
            var card = SpacedRepetition.createCard('test');
            var updated = SpacedRepetition.review(card, 1);
            assert.ok(updated.ease_factor < card.ease_factor);
        });

        it('adjusts ease factor upward on easy answer', function () {
            var card = SpacedRepetition.createCard('test');
            var updated = SpacedRepetition.review(card, 5);
            assert.ok(updated.ease_factor > card.ease_factor);
        });

        it('never lets ease factor drop below MIN_EASE', function () {
            var card = SpacedRepetition.createCard('test');
            // Repeatedly get wrong answers to drive EF down
            for (var i = 0; i < 20; i++) {
                card = SpacedRepetition.review(card, 1);
            }
            assert.ok(card.ease_factor >= SpacedRepetition.MIN_EASE);
        });

        it('sets next_review in the future', function () {
            var card = SpacedRepetition.createCard('test');
            var before = Date.now();
            var updated = SpacedRepetition.review(card, 4);
            var nextReview = new Date(updated.next_review).getTime();
            assert.ok(nextReview >= before);
        });

        it('records last_quality', function () {
            var card = SpacedRepetition.createCard('test');
            var updated = SpacedRepetition.review(card, 3);
            assert.equal(updated.last_quality, 3);
        });

        it('throws on invalid quality value', function () {
            var card = SpacedRepetition.createCard('test');
            assert.throws(function () {
                SpacedRepetition.review(card, 0);
            }, /quality must be 1, 3, 4, or 5/);
            assert.throws(function () {
                SpacedRepetition.review(card, 2);
            }, /quality must be 1, 3, 4, or 5/);
            assert.throws(function () {
                SpacedRepetition.review(card, 6);
            }, /quality must be 1, 3, 4, or 5/);
        });
    });

    describe('isDue', function () {
        it('returns true for a new card', function () {
            var card = SpacedRepetition.createCard('test');
            assert.ok(SpacedRepetition.isDue(card));
        });

        it('returns false for a recently reviewed card', function () {
            var card = SpacedRepetition.createCard('test');
            var updated = SpacedRepetition.review(card, 4);
            assert.equal(SpacedRepetition.isDue(updated), false);
        });
    });

    describe('overdueMinutes', function () {
        it('returns positive for overdue cards', function () {
            var card = SpacedRepetition.createCard('test');
            // A new card is due immediately, so should be >= 0
            var overdue = SpacedRepetition.overdueMinutes(card);
            assert.ok(overdue >= 0);
        });

        it('returns negative for cards not yet due', function () {
            var card = SpacedRepetition.createCard('test');
            var updated = SpacedRepetition.review(card, 4);
            var overdue = SpacedRepetition.overdueMinutes(updated);
            assert.ok(overdue < 0);
        });
    });
});
